core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -141,7 +141,14 @@ final class ShuffleBlockFetcherIterator(
 
   /**
    * Whether the iterator is still active. If isZombie is true, the callback interface will no
-   * longer place fetched blocks into [[results]].
+   * longer place fetched blocks into [[results]] and the iterator is marked as fully consumed.
+   *
+   * When the iterator is inactive, [[hasNext]] and [[next]] calls will honor that as there are
+   * cases the iterator is still being consumed. For example, ShuffledRDD + PipedRDD if the
+   * subprocess command is failed. The task will be marked as failed, then the iterator will be
+   * cleaned up at task completion, the [[next]] call (called in the stdin writer thread of
+   * PipedRDD if not exited yet) may hang at [[results.take]]. The defensive check in [[hasNext]]
+   * and [[next]] reduces the possibility of such race conditions.
```

Comment:
When a task finishes, do we really need to guarantee all the iterators stop producing data? I agree it's better, but I'm afraid it's too much effort to guarantee it. Not only shuffle reader, we also need to fix the sort iterator, aggregate iterator and so on.

And why can't `PipelinedRDD` stop consuming input? I think it's better to fix the solo consumer side, instead of fixing different kinds of producer sides.

Reply:
"try best" is not a "guarantee".

If we don't need to do this, I suggest we should not do it at all. The new changes in `ShuffleBlockFetcherIterator` make it harder for people to understand the code(at least to me), and also breaks the semantic of Iterator. And I don't see much benefit of doing it, as the `PipedRDD` has been fixed. Can we revert the changes in `ShuffleBlockFetcherIterator`?

Reply:
> Can we revert the changes in ShuffleBlockFetcherIterator?

If you insist, I can revert those changes. But let's wait and see if others have other opinions. 
cc @srowen, @HyukjinKwon and @viirya.

> also breaks the semantic of Iterator.

Rarely and shouldn't matter that much if the task is already finished.

Reply:
I agree that this is not the expected behavior of Iterator. If there are elements in Iterator, it should return true when `hasNext` is called. It sounds more reasonable at the consumer side to stop consuming Iterator.

Reply:
> If there are elements in Iterator, it should return true when hasNext is called.

The problem is that the old behaviour is not as expected as well. `hasNext` called and returns `true`,  then `next` called, it may hangs forever.

Reply:
>  then next called, it may hangs forever.

can you elaborate a little more about it? If it's true then this is a critical bug, which may affect all the queries with shuffles. I thought the hang problem only happens in `PipedRDD`.

Reply:
Ok I got it. The thing we need to guarantee is, the iterator should not hang in any cases, even if the task has finished and cleanup is called.

`ShuffleBlockFetcherIterator` does have a bug, that there can be 2 threads consuming the `results` queue at the same time, and cause `next` to hang.

I don't think the fix here is corrected. Right after `while (!isZombie && result == null)`, the other thread may set `isZombie` to `true` and consume up the `result` queue, and we may still hang here. It does reduce the possibility of this race condition, but the problem is still there.

Can we revert it and come up with a better fix? I think we should do the heavy work in `hasNext` and put the result in a temporary variable, then the `next` just returns that temporary variable.

Reply:
Sure, how about a follow-up that tries a different approach? the current change isn't harmful per se, and a small improvement.

You're suggesting reading and storing the next element that's available, if not already read, in `hasNext`? and then `next` must call `hasNext` to ensure this is filled if not already, which it does already? yeah that seems reasonable. That pattern is used in other iterators sometimes.

Reply:
> how about a follow-up that tries a different approach

Sure, if we can do it soon. Hopefully we don't leave this partial fix in the code base for years...

Reply:
@advancedxy thanks for working on it! I'm leaving for lunar new year soon too, end of Feb sounds good.

Reply:
@advancedxy  I am facing the the same issue discussed by @cloud-fan  that there are 2 threads consuming the results queue at the same time, and causing spark application to hang. Is the fix for this issue being worked on right now . Is there a JIRA to track the fix?